### PR TITLE
Add docker container.

### PR DESCRIPTION
**Pull request by larroy**
Added docker container that gets the ibgateway from ib, and builds from source, so it's reasonably trustworthy.

### CODE REVIEW

**Comment by rlktradewright** [comment]
I'm not convinced that it's appropriate for this repository to include any sort of 'recommended' docker image. Certainly not right up in the root folder.

There are many ways to deploy IBC, and many different use cases. It is really a matter for individual IBC users to work out how to deploy IBC to meet their needs (other than the obvious 'normal' case of just running the supplied scripts).

However, you can make a case for a sample docker setup just to show the basics of how it might be done, and in that case it should be under the Samples folder. If you do this, then please note the following:

* Provide enough documentation that a Docker newbie can make it work - so give it a README in the relevant folder (NOT the IBC User Guide). In particular, before accepting it I will need to follow your instructions and it should 'just work' (I'll do this on Ubuntu 18.04 LTS)

* Be explicit about which Linux distributions it will work with (not a detailed list of distributions, obviously, but generically - eg if it needs systemd, say so)

* Be prepared to respond promptly to any support requests relating to it: I am not qualified to do so

* Take responsibility for maintaining it - if it falls into disrepair, I'll just remove it

**Comment by larroy** [comment]
Ok, thanks for your feedback, will do so.

**Comment by rlktradewright** [comment]
@larroy

Are you waiting for further response from me, or have you decided not to proceed with this, or just haven't had the time yet?

**Comment by larroy** [comment]
didn't have the time yet.

**Comment by larroy** [comment]
The docker compose needs to be in the root folder to have access to the project to build, the rest of the files are in the docker folder. Is this ok to merge?  I revisited the documentation.

**Comment by rlktradewright** [comment]
@larroy 

First let me apologise for taking so long to respond to this. I wanted to test out your docker in both
Linux and Windows, and I finally got round to doing it a couple of days ago. To cut a long story
short, I eventually got it working in both, and it was very useful experience for me as I haven't
given Docker much attention in the past, primarily because I have no real use for it myself and I simply have too many other higher priority things to do.

As a result of this I have a number of comments to make.

1. No, I won't accept a docker-compose.yml in the root folder. As I said, if there's going to be a
   docker at all, it's going to be in the samples/docker folder. I managed to modify your files to
   work from this folder, and actually to be independent of the repository altogether (ie a free-
   standing dockerfile build that anyone could run, whether or not they have cloned the repository).
   I did this by using:

   `RUN git clone -q https://github.com/IbcAlpha/IBC.git`

   rather than assuming you're in the root of a clone. This of course will always download the latest
   files, and there are situations where that may not be appropriate, but this is only supposed to
   be a simple demonstration of one way to set up a docker image, and people can modify it to suit
   their needs.

2. Setting the TWSUSERID and TWSPASSWORD environment variables in docker-compose.yml is 
   pointless, since they are not used anywhere.

3. Your changes to build.xml, and the line in Dockerfile that creates the lib folder are unnecessary:
   the project builds fine without these.

4. You didn't actually include any documentation. The changes you made to userguide.md refer to
   `[docker](docker/README.md)` but this file doesn't actually exist.

5. Please remove the changes to userguide.md: I haven't yet decided whether I want to mention
   Docker at all in the User Guide, so I'll make that decision myself and provide any relevant updates.

6. Your files as provided don't actually successfully build the Docker image: they assume the
   presence of `config.ini` in the root folder, and that's not where it is. So it needs to be in the
   `samples/docker` folder, and there needs to be a comment somewhere that the user should set
   the username and password, and any other required settings, in this file.

7. In docker_compose.yml, ibc_tws is perhaps not an ideal name, since this Docker image is for the
   Gateway, not TWS.

8. In supervisord.conf, replace `[program:ibcontroller]` with `[program:ibc]`.

9. It's not clear to me what determines the IP address of the running Gateway, and thus how to
   connect to the API. Please document.

10. The link to https://xpra.org/Xdummy.html in xorg.conf is stale.

11. I presume I should be able to access the Gateway gui using VNC or similar? If only I knew the IP
   address...

12. It's arguable whether building the IBC project like this is the best way to proceed. Might it be
   better to target the source code from the latest, or even a specific, release (eg download and
   expand the tar.gz file) and build from that, or even just download the IBC.jar file directly and
   not bother with the build? This would avoid potential problems with the repository not being in a
   clean state (for example a file has been pushed that contains bugs, but has not yet been included
   in a release). I realise this sort of thing shouldn't happen, but we need to be realistic.

Ok, I think that will do for now. Keep up the good work!

**Comment by rlktradewright** [comment]
Oh yes, another point:

13. You don't need to supply a jts.ini: one will be created automatically.

**Comment by larroy** [comment]
Without my changes to build.xml I get the following build failure:


```
Buildfile: /work/ibc/build.xml

init:

compile:
    [mkdir] Created dir: /work/ibc/target/classes
    [javac] Compiling 65 source files to /work/ibc/target/classes
    [javac] /work/ibc/src/ibcalpha/ibc/DefaultMainWindowManager.java:32: error: package javafx.application does not exist
    [javac] import javafx.application.Preloader;
    [javac]                          ^
    [javac] 1 error


```

This is the part that builds the library:

```
RUN mkdir lib && cp /usr/local/i4j_jres/1.8.0_152-tzdata2019c/lib/ext/jfxrt.jar lib
RUN IBC_BIN=/root/Jts/ibgateway/972/jars/ ant
```

Any suggestions?  Added readme and restored userguide as requested.
I'm going through your requests.

**Comment by rlktradewright** [comment]
This compilation error might be because of an unnecessary import in `DefaultMainWindowManager.java`, which was removed in commit 66f8a7f211fd77f519cc1edf892cf3cf4f03b0f9 on 1 April.

Perhaps you need to pull from your upstream to ensure you have all the latest commits?

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,35 @@
+FROM ubuntu:18.04
```
Comment (peterfig) [inline docker/Dockerfile:1]: Would it be possible to use a lightweight linux image such as alpine, instead of ubuntu?

**Comment by dvasdekis** [comment]
I maintain an updated docker container with IB Gateway v978.2c and IBC v3.8.2 [here](https://github.com/dvasdekis/ib-gateway-docker-gcp). Feel free to link to it instead of maintaining an independent solution.

**Comment by rlktradewright** [comment]
@larroy 

I'm closing this because there as been no further input from you. If you want to resurrect it, feel free, but if so it needs to be moved forward.

**Pull request closed without merging**